tools/topology/src/symbol.rs

```diff
@@ -0,0 +1,106 @@
+use sha2::{Digest as _, Sha256};
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct Symbol {
+    path: Box<str>,
+    symbol: Box<str>,
+}
+
+impl Symbol {
+    pub fn path(&self) -> &str {
+        &self.path
+    }
+
+    pub fn symbol(&self) -> &str {
+        &self.symbol
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct Builder(Inner);
+
+impl Builder {
+    // TODO replace with following after upgrade to Rust 1.79+:
+    //  ```
+    //  pub const fn new() -> Self {
+    //      const { Self(Inner::Native) }
+    //  }
+    pub const NEW: Self = Self(Inner::Native);
+
+    pub fn add_channel(&mut self, channel: &str) {
+        match &mut self.0 {
+            Inner::Native => self.0 = Inner::Ibc(Ibc::new(channel)),
+            Inner::Ibc(ibc_symbol) => ibc_symbol.add_channel(channel),
+        }
+    }
+
+    pub fn add_symbol(self, symbol: &str) -> Symbol {
+        match self.0 {
+            Inner::Native => Symbol {
+                path: symbol.into(),
+                symbol: symbol.into(),
+            },
+            Inner::Ibc(ibc_symbol) => ibc_symbol.add_symbol(symbol),
+        }
+    }
+}
+
+#[derive(Debug)]
+enum Inner {
+    Native,
+    Ibc(Ibc),
+}
+
+#[derive(Debug)]
+struct Ibc {
+    path: String,
+}
+
+impl Ibc {
+    fn new(outmost_channel: &str) -> Self {
+        let mut instance = Self {
+            path: String::new(),
+        };
+
+        instance.add_channel(outmost_channel);
+
+        instance
+    }
+
+    fn add_channel(&mut self, channel: &str) {
+        self.path.push_str("transfer/");
+
+        self.path.push_str(channel);
+
+        self.path.push('/');
+    }
+
+    fn add_symbol(mut self, symbol: &str) -> Symbol {
+        self.path.push_str(symbol);
+
+        // "ibc/" + 64 hexadecimal digits
+        let mut symbol = String::new();
+
+        symbol.reserve_exact(68);
+
+        symbol.push_str("ibc/");
+
+        Sha256::digest(&self.path)
+            .into_iter()
+            .flat_map(|byte| [byte >> 4, byte & 0xF])
+            .map(|half_byte| {
+                if half_byte < 10 {
+                    b'0' + half_byte
+                } else {
+                    b'A' + (half_byte - 10)
+                }
+            })
+            .map(char::from)
+            .for_each(|ch| symbol.push(ch));
+
+        Symbol {
+            path: self.path.into_boxed_str(),
+            symbol: symbol.into_boxed_str(),
```

Comment:
Why do we complicate the stuff working with `Box<str>` when we own the strings? In addition, `into_boxed_str` may reallocate and copy the raw bytes.

Reply:
It drops the excess capacity. While yes, it could reallocate, we won't be leaking memory that won't ever be needed as we won't even need to modify the string.

Please take take a look at this video: https://youtu.be/A4cKi7PTJSs. I'm not using `Rc<str>` or `Arc<str>` since probably it would be an overkill and I know your opinion on it, but `Box<str>` is sufficient. We don't clone this in the crate, nor do we modify it, so it's a nice compromise between the two.

Reply:
simplicity is always preferred than obscurity ... and I can send you 10x materials on that matter

Reply:
What you are saying is that if we were to use e.g. Java, you'd leave the whole application use `StringBuilder`. I hope you realize how fanatical that sounds.  
That is the literal definition of using the wrong tool for the job and I'm dying on this hill.

Would a type alias satisfy your requirement for simplicity and readability?